File: examples/rdma/go/s3rdmamultipart.go
//go:build linux

package main

import (
	"bytes"
	"context"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"syscall"
	"time"
	"unsafe"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/gorilla/websocket"
)

const (
	multipartFlagPort   = 1
	multipartFlagGIDIdx = 0
)

type rdmaPartHeaderTransport struct {
	base      http.RoundTripper
	requestID string
}

func (t *rdmaPartHeaderTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	req.Header.Set("X-RDMA-Request-ID", t.requestID)
	// For RDMA UploadPart, the HTTP body is empty. The actual part size is in the RDMA token.
	req.ContentLength = 0
	return t.base.RoundTrip(req)
}

func main() {
	bucket := flag.String("bucket", "", "Name of the bucket")
	objectKey := flag.String("object", "", "Name of the object")
	filePath := flag.String("file", "", "Path to the file to upload")
	partSize := flag.Int64("partsize", 5*1024*1024, "Size of each multipart part in bytes")
	action := flag.String("action", "all", "Action to perform: all, create, upload, list, list-uploads, complete, abort")
	uploadID := flag.String("uploadid", "", "Upload ID for multipart operations")
	partNumber := flag.Int("part", 0, "Part number for upload action; 0 uploads all parts")
	prefix := flag.String("prefix", "", "Prefix for filtering results (for list-uploads action)")
	delimiter := flag.String("delimiter", "", "Delimiter for grouping keys (for list-uploads action)")
	maxUploads := flag.Int64("max-uploads", 1000, "Maximum number of uploads to return (for list-uploads action)")
	keyMarker := flag.String("key-marker", "", "Key marker for pagination (for list-uploads action)")
	uploadIDMarker := flag.String("upload-id-marker", "", "Upload ID marker for pagination (for list-uploads action)")
	endpoint := flag.String("endpoint", "http://127.0.0.1:8901", "S3 endpoint URL")
	rdmaDev := flag.String("rdma-dev", "rxe0", "RDMA device name")
	ak := flag.String("ak", "admin-ak", "S3 access key")
	sk := flag.String("sk", "admin-sk", "S3 secret key")
	flag.Parse()

	if *bucket == "" {
		log.Fatal("Usage: go run s3rdmamultipart.go rdmalib.go -bucket <bucketname> -action <action> [options]")
	}
	if *action != "list-uploads" && *objectKey == "" {
		log.Fatal("object is required for this action")
	}
	if *partSize <= 0 {
		log.Fatal("partsize must be greater than 0")
	}

	ctx := context.Background()
	cfg := newS3Config(*endpoint, *ak, *sk)
	client := newS3Client(cfg)

	switch *action {
	case "all":
		if *filePath == "" {
			log.Fatal("Usage: -action all requires -file")
		}
		fileInfo := mustStatUploadFile(*filePath)
		if fileInfo.Size() == 0 {
			log.Fatal("multipart RDMA upload does not support empty files")
		}
		newUploadID := createMultipartUpload(ctx, client, *bucket, *objectKey)
		completedParts, err := uploadFilePartsRDMA(ctx, cfg, *endpoint, *rdmaDev, *bucket, *objectKey, *filePath, newUploadID, fileInfo.Size(), *partSize, 0)
		if err != nil {
			log.Printf("Upload failed, aborting multipart upload: %v", err)
			abortMultipartUpload(ctx, client, *bucket, *objectKey, newUploadID)
			log.Fatal(err)
		}
		completeMultipartUpload(ctx, client, *bucket, *objectKey, newUploadID, completedParts)
		log.Printf("Successfully uploaded %s to s3://%s/%s via RDMA multipart", *filePath, *bucket, *objectKey)

	case "create":
		createMultipartUpload(ctx, client, *bucket, *objectKey)

	case "upload":
		if *filePath == "" || *uploadID == "" {
			log.Fatal("Usage: -action upload requires -file and -uploadid; use -part <n> for one part or omit/0 for all parts")
		}
		fileInfo := mustStatUploadFile(*filePath)
		if fileInfo.Size() == 0 {
			log.Fatal("multipart RDMA upload does not support empty files")
		}
		completedParts, err := uploadFilePartsRDMA(ctx, cfg, *endpoint, *rdmaDev, *bucket, *objectKey, *filePath, *uploadID, fileInfo.Size(), *partSize, *partNumber)
		if err != nil {
			log.Fatal(err)
		}
		for _, part := range completedParts {
			fmt.Printf("Part %d ETag: %s\n", aws.ToInt32(part.PartNumber), aws.ToString(part.ETag))
		}

	case "list":
		if *uploadID == "" {
			log.Fatal("Usage: -action list requires -uploadid")
		}
		listParts(ctx, client, *bucket, *objectKey, *uploadID)

	case "list-uploads":
		listMultipartUploads(ctx, client, *bucket, *prefix, *delimiter, *maxUploads, *keyMarker, *uploadIDMarker)

	case "complete":
		if *uploadID == "" {
			log.Fatal("Usage: -action complete requires -uploadid")
		}
		parts, err := getCompletedParts(ctx, client, *bucket, *objectKey, *uploadID)
		if err != nil {
			log.Fatal(err)
		}
		completeMultipartUpload(ctx, client, *bucket, *objectKey, *uploadID, parts)

	case "abort":
		if *uploadID == "" {
			log.Fatal("Usage: -action abort requires -uploadid")
		}
		abortMultipartUpload(ctx, client, *bucket, *objectKey, *uploadID)

	default:
		log.Fatalf("Unknown action: %s. Valid actions are: all, create, upload, list, list-uploads, complete, abort", *action)
	}
}

func newS3Config(endpoint, ak, sk string) aws.Config {
	return aws.Config{
		Region:      "us-east-1",
		Credentials: credentials.NewStaticCredentialsProvider(ak, sk, ""),
		EndpointResolver: aws.EndpointResolverFunc(func(service, region string) (aws.Endpoint, error) {
			return aws.Endpoint{
				URL:               endpoint,
				SigningRegion:     "us-east-1",
				HostnameImmutable: true,
			}, nil
		}),
	}
}

func newS3Client(cfg aws.Config) *s3.Client {
	return s3.NewFromConfig(cfg, func(o *s3.Options) {
		o.UsePathStyle = true
	})
}

func mustStatUploadFile(filePath string) os.FileInfo {
	fileInfo, err := os.Stat(filePath)
	if err != nil {
		log.Fatalf("Failed to stat file: %v", err)
	}
	return fileInfo
}

func createMultipartUpload(ctx context.Context, client *s3.Client, bucket, objectKey string) string {
	resp, err := client.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(objectKey),
	})
	if err != nil {
		log.Fatalf("CreateMultipartUpload error: %v", err)
	}
	uploadID := aws.ToString(resp.UploadId)
	fmt.Printf("Successfully created multipart upload\n")
	fmt.Printf("Upload ID: %s\n", uploadID)
	return uploadID
}

func uploadFilePartsRDMA(ctx context.Context, cfg aws.Config, endpoint, rdmaDev, bucket, objectKey, filePath, uploadID string, fileSize, partSize int64, onlyPart int) ([]types.CompletedPart, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("open file: %w", err)
	}
	defer file.Close()

	numParts := int((fileSize + partSize - 1) / partSize)
	if onlyPart < 0 || onlyPart > numParts {
		return nil, fmt.Errorf("part must be between 1 and %d, or 0 for all parts", numParts)
	}

	startPart := 1
	endPart := numParts
	if onlyPart > 0 {
		startPart = onlyPart
		endPart = onlyPart
	}

	completedParts := make([]types.CompletedPart, 0, endPart-startPart+1)
	for partNumber := startPart; partNumber <= endPart; partNumber++ {
		offset := int64(partNumber-1) * partSize
		currentPartSize := partSize
		if remaining := fileSize - offset; remaining < currentPartSize {
			currentPartSize = remaining
		}

		etag, err := uploadPartRDMA(ctx, cfg, endpoint, rdmaDev, file, bucket, objectKey, uploadID, partNumber, offset, currentPartSize)
		if err != nil {
			return nil, fmt.Errorf("upload part %d: %w", partNumber, err)
		}

		pn := int32(partNumber)
		completedParts = append(completedParts, types.CompletedPart{
			ETag:       aws.String(etag),
			PartNumber: &pn,
		})
		log.Printf("Uploaded part %d/%d, size=%d, etag=%s", partNumber, numParts, currentPartSize, etag)
	}

	return completedParts, nil
}

func uploadPartRDMA(ctx context.Context, cfg aws.Config, endpoint, rdmaDev string, file *os.File, bucket, objectKey, uploadID string, partNumber int, offset, partSize int64) (string, error) {
	requestID := fmt.Sprintf("rdma-multipart-%d-part-%d-%d", os.Getpid(), partNumber, time.Now().UnixNano())
	buf, err := mmapPart(file, offset, partSize)
	if err != nil {
		return "", err
	}
	defer syscall.Munmap(buf)

	wsConn, err := connectRDMAControl(endpoint, requestID)
	if err != nil {
		return "", err
	}
	defer wsConn.Close()

	ctxRDMA, err := OpenDevice(rdmaDev)
	if err != nil {
		return "", fmt.Errorf("open RDMA device: %w", err)
	}
	defer ctxRDMA.Close()

	pd, err := ctxRDMA.AllocPD()
	if err != nil {
		return "", fmt.Errorf("alloc PD: %w", err)
	}
	defer pd.Dealloc()

	cq, err := ctxRDMA.CreateCQ(2)
	if err != nil {
		return "", fmt.Errorf("create CQ: %w", err)
	}
	defer cq.Destroy()

	qp, err := pd.CreateRC(cq, 1, 1, 0)
	if err != nil {
		return "", fmt.Errorf("create QP: %w", err)
	}
	defer qp.Destroy()

	if err := qp.ToInit(multipartFlagPort); err != nil {
		return "", fmt.Errorf("set QP INIT: %w", err)
	}

	gid, err := ctxRDMA.QueryGID(multipartFlagPort, multipartFlagGIDIdx)
	if err != nil {
		return "", fmt.Errorf("query GID: %w", err)
	}
	mtuEnum, err := ctxRDMA.QueryMTUEnum(multipartFlagPort)
	if err != nil {
		return "", fmt.Errorf("query MTU: %w", err)
	}

	clientQPInfo := QPInfo{
		QPN:     qp.Num,
		LID:     0,
		GID:     gid,
		GIDIdx:  multipartFlagGIDIdx,
		MTUEnum: int32(mtuEnum),
	}
	if err := sendJSON(wsConn, MsgTypeQPInfoClient, clientQPInfo); err != nil {
		return "", fmt.Errorf("send QP info: %w", err)
	}

	var serverQPInfo QPInfo
	if err := recvJSON(wsConn, MsgTypeQPInfoServer, &serverQPInfo); err != nil {
		return "", fmt.Errorf("receive server QP info: %w", err)
	}

	if err := qp.ToRTR(serverQPInfo.QPN, serverQPInfo.LID, serverQPInfo.GID, multipartFlagPort, serverQPInfo.GIDIdx, int(serverQPInfo.MTUEnum)); err != nil {
		return "", fmt.Errorf("set QP RTR: %w", err)
	}
	if err := qp.ToRTS(); err != nil {
		return "", fmt.Errorf("set QP RTS: %w", err)
	}

	mr, err := pd.RegMRSrc(unsafe.Pointer(&buf[0]), uint64(partSize))
	if err != nil {
		return "", fmt.Errorf("register source MR: %w", err)
	}
	defer mr.Dereg()

	token := Token{
		Addr:   uint64(uintptr(unsafe.Pointer(&buf[0]))),
		Rkey:   mr.Rkey,
		Start:  0,
		Length: uint64(partSize),
	}
	if err := sendJSON(wsConn, MsgTypeToken, token); err != nil {
		return "", fmt.Errorf("send token: %w", err)
	}

	if err := closeControlAfterToken(wsConn); err != nil {
		log.Printf("Warning: could not close RDMA control connection cleanly: %v", err)
	}

	rdmaClient := s3.NewFromConfig(cfg, func(o *s3.Options) {
		o.UsePathStyle = true
		o.HTTPClient = &http.Client{
			Transport: &rdmaPartHeaderTransport{
				base:      http.DefaultTransport,
				requestID: requestID,
			},
		}
	})

	pn := int32(partNumber)
	resp, err := rdmaClient.UploadPart(ctx, &s3.UploadPartInput{
		Bucket:        aws.String(bucket),
		Key:           aws.String(objectKey),
		UploadId:      aws.String(uploadID),
		PartNumber:    &pn,
		Body:          bytes.NewReader(nil),
		ContentLength: aws.Int64(0),
	})
	if err != nil {
		return "", fmt.Errorf("UploadPart request: %w", err)
	}

	return aws.ToString(resp.ETag), nil
}

func listParts(ctx context.Context, client *s3.Client, bucket, objectKey, uploadID string) {
	resp, err := client.ListParts(ctx, &s3.ListPartsInput{
		Bucket:   aws.String(bucket),
		Key:      aws.String(objectKey),
		UploadId: aws.String(uploadID),
	})
	if err != nil {
		log.Fatalf("ListParts error: %v", err)
	}

	fmt.Printf("Parts uploaded for upload ID %s:\n", aws.ToString(resp.UploadId))
	for _, part := range resp.Parts {
		fmt.Printf("Part %d: ETag=%s, Size=%d bytes\n", aws.ToInt32(part.PartNumber), aws.ToString(part.ETag), aws.ToInt64(part.Size))
	}
}

func getCompletedParts(ctx context.Context, client *s3.Client, bucket, objectKey, uploadID string) ([]types.CompletedPart, error) {
	resp, err := client.ListParts(ctx, &s3.ListPartsInput{
		Bucket:   aws.String(bucket),
		Key:      aws.String(objectKey),
		UploadId: aws.String(uploadID),
	})
	if err != nil {
		return nil, fmt.Errorf("ListParts error: %w", err)
	}
	if len(resp.Parts) == 0 {
		return nil, fmt.Errorf("no uploaded parts found for uploadId %s", uploadID)
	}

	parts := make([]types.CompletedPart, 0, len(resp.Parts))
	for _, part := range resp.Parts {
		partNumber := aws.ToInt32(part.PartNumber)
		parts = append(parts, types.CompletedPart{
			ETag:       aws.String(aws.ToString(part.ETag)),
			PartNumber: aws.Int32(partNumber),
		})
	}
	return parts, nil
}

func listMultipartUploads(ctx context.Context, client *s3.Client, bucket, prefix, delimiter string, maxUploads int64, keyMarker, uploadIDMarker string) {
	maxUploads32 := int32(maxUploads)
	resp, err := client.ListMultipartUploads(ctx, &s3.ListMultipartUploadsInput{
		Bucket:         aws.String(bucket),
		Prefix:         aws.String(prefix),
		Delimiter:      aws.String(delimiter),
		MaxUploads:     &maxUploads32,
		KeyMarker:      aws.String(keyMarker),
		UploadIdMarker: aws.String(uploadIDMarker),
	})
	if err != nil {
		log.Fatalf("ListMultipartUploads error: %v", err)
	}

	fmt.Printf("Multipart uploads for bucket: %s\n", aws.ToString(resp.Bucket))
	if resp.Prefix != nil {
		fmt.Printf("Prefix: %s\n", aws.ToString(resp.Prefix))
	}
	if resp.Delimiter != nil {
		fmt.Printf("Delimiter: %s\n", aws.ToString(resp.Delimiter))
	}
	fmt.Printf("MaxUploads: %d\n", aws.ToInt32(resp.MaxUploads))
	fmt.Printf("IsTruncated: %v\n", resp.IsTruncated)
	if resp.KeyMarker != nil {
		fmt.Printf("KeyMarker: %s\n", aws.ToString(resp.KeyMarker))
	}
	if resp.UploadIdMarker != nil {
		fmt.Printf("UploadIdMarker: %s\n", aws.ToString(resp.UploadIdMarker))
	}
	if aws.ToBool(resp.IsTruncated) {
		if resp.NextKeyMarker != nil {
			fmt.Printf("NextKeyMarker: %s\n", aws.ToString(resp.NextKeyMarker))
		}
		if resp.NextUploadIdMarker != nil {
			fmt.Printf("NextUploadIdMarker: %s\n", aws.ToString(resp.NextUploadIdMarker))
		}
	}
	fmt.Printf("\nNumber of uploads: %d\n", len(resp.Uploads))

	for i, upload := range resp.Uploads {
		fmt.Printf("\nUpload %d:\n", i+1)
		fmt.Printf("  Key: %s\n", aws.ToString(upload.Key))
		fmt.Printf("  UploadId: %s\n", aws.ToString(upload.UploadId))
		fmt.Printf("  StorageClass: %s\n", string(upload.StorageClass))
		if upload.Initiated != nil {
			fmt.Printf("  Initiated: %v\n", upload.Initiated.Format(time.RFC3339))
		}
	}

	if len(resp.CommonPrefixes) > 0 {
		fmt.Printf("\nNumber of common prefixes: %d\n", len(resp.CommonPrefixes))
		for i, prefix := range resp.CommonPrefixes {
			fmt.Printf("\nCommon Prefix %d:\n", i+1)
			fmt.Printf("  Prefix: %s\n", aws.ToString(prefix.Prefix))
		}
	}
}

func completeMultipartUpload(ctx context.Context, client *s3.Client, bucket, objectKey, uploadID string, parts []types.CompletedPart) {
	resp, err := client.CompleteMultipartUpload(ctx, &s3.CompleteMultipartUploadInput{
		Bucket:   aws.String(bucket),
		Key:      aws.String(objectKey),
		UploadId: aws.String(uploadID),
		MultipartUpload: &types.CompletedMultipartUpload{
			Parts: parts,
		},
	})
	if err != nil {
		log.Fatalf("CompleteMultipartUpload error: %v", err)
	}

	fmt.Printf("Successfully completed multipart upload\n")
	fmt.Printf("Location: %s\n", aws.ToString(resp.Location))
	fmt.Printf("ETag: %s\n", aws.ToString(resp.ETag))
	fmt.Printf("Bucket: %s\n", aws.ToString(resp.Bucket))
	fmt.Printf("Key: %s\n", aws.ToString(resp.Key))
}

func abortMultipartUpload(ctx context.Context, client *s3.Client, bucket, objectKey, uploadID string) {
	_, err := client.AbortMultipartUpload(ctx, &s3.AbortMultipartUploadInput{
		Bucket:   aws.String(bucket),
		Key:      aws.String(objectKey),
		UploadId: aws.String(uploadID),
	})
	if err != nil {
		log.Fatalf("AbortMultipartUpload error: %v", err)
	}
	fmt.Printf("Successfully aborted multipart upload\n")
}

func mmapPart(file *os.File, offset, partSize int64) ([]byte, error) {
	if partSize <= 0 {
		return nil, fmt.Errorf("part size must be greater than 0")
	}
	if int64(int(partSize)) != partSize {
		return nil, fmt.Errorf("part size %d is too large for mmap length", partSize)
	}

	buf, err := syscall.Mmap(-1, 0, int(partSize), syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_PRIVATE|syscall.MAP_ANON)
	if err != nil {
		return nil, fmt.Errorf("mmap: %w", err)
	}

	n, err := file.ReadAt(buf, offset)
	if err != nil && err != io.EOF {
		syscall.Munmap(buf)
		return nil, fmt.Errorf("read file part: %w", err)
	}
	if int64(n) != partSize {
		syscall.Munmap(buf)
		return nil, fmt.Errorf("read %d bytes, expected %d", n, partSize)
	}

	return buf, nil
}

func connectRDMAControl(endpoint, requestID string) (*websocket.Conn, error) {
	wsURL, err := url.Parse(endpoint)
	if err != nil {
		return nil, fmt.Errorf("parse endpoint: %w", err)
	}
	if wsURL.Scheme == "https" {
		wsURL.Scheme = "wss"
	} else {
		wsURL.Scheme = "ws"
	}
	wsURL.Path = "/rdma-ctrl"
	wsURL.RawQuery = ""

	wsConn, _, err := websocket.DefaultDialer.Dial(wsURL.String(), nil)
	if err != nil {
		return nil, fmt.Errorf("connect WebSocket: %w", err)
	}

	if err := sendJSON(wsConn, MsgTypeRegisterRequest, RegisterRequest{RequestID: requestID}); err != nil {
		wsConn.Close()
		return nil, fmt.Errorf("send register request: %w", err)
	}

	var regResp RegisterResponse
	if err := recvJSON(wsConn, MsgTypeRegisterResponse, &regResp); err != nil {
		wsConn.Close()
		return nil, fmt.Errorf("receive register response: %w", err)
	}
	if !regResp.Success {
		wsConn.Close()
		return nil, fmt.Errorf("registration failed: %s", regResp.Error)
	}

	return wsConn, nil
}

func closeControlAfterToken(wsConn *websocket.Conn) error {
	return wsConn.WriteControl(
		websocket.CloseMessage,
		websocket.FormatCloseMessage(websocket.CloseNormalClosure, "Token sent, closing early"),
		time.Now().Add(time.Second),
	)
}

File: examples/rdma/python/s3rdmamultipart.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""s3rdmamultipart.py - multipart upload to S3 via RDMA.
Python 2.7 / 3.x compatible.
"""
from __future__ import print_function

import argparse
import hashlib
import hmac
import os
import sys
import time
try:
    from urllib.parse import urlsplit
except ImportError:
    from urlparse import urlsplit
try:
    import http.client as http_client
except ImportError:
    import httplib as http_client
try:
    import xml.etree.ElementTree as ET
except ImportError:
    from xml.etree import ElementTree as ET

import rdmalib

UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD"


class S3Config(object):
    def __init__(self, endpoint, region, access_key, secret_key):
        self.endpoint = endpoint
        self.region = region
        self.access_key = access_key
        self.secret_key = secret_key


def _to_bytes(s):
    if isinstance(s, bytes):
        return s
    return s.encode("utf-8")


def _hmac_sha256(key, msg):
    return hmac.new(key, _to_bytes(msg), hashlib.sha256).digest()


def _sha256_hex(s):
    return hashlib.sha256(_to_bytes(s)).hexdigest()


def _derive_signing_key(secret_key, datestamp, region):
    k_date = _hmac_sha256(_to_bytes("AWS4" + secret_key), datestamp)
    k_region = _hmac_sha256(k_date, region)
    k_service = _hmac_sha256(k_region, "s3")
    return _hmac_sha256(k_service, "aws4_request")


def _sign_request(cfg, method, canonical_uri, canonical_query, host, amzdate, datestamp, content_length=None):
    if content_length is None:
        canonical_headers = "host:%s\nx-amz-date:%s\n" % (host, amzdate)
        signed_headers = "host;x-amz-date"
    else:
        canonical_headers = "content-length:%d\nhost:%s\nx-amz-date:%s\n" % (content_length, host, amzdate)
        signed_headers = "content-length;host;x-amz-date"

    canonical_request = "%s\n%s\n%s\n%s\n%s\n%s" % (
        method, canonical_uri, canonical_query or "", canonical_headers, signed_headers, UNSIGNED_PAYLOAD)
    creq_hash = _sha256_hex(canonical_request)

    scope = "%s/%s/s3/aws4_request" % (datestamp, cfg.region)
    string_to_sign = "AWS4-HMAC-SHA256\n%s\n%s\n%s" % (amzdate, scope, creq_hash)
    signing_key = _derive_signing_key(cfg.secret_key, datestamp, cfg.region)
    signature = hmac.new(signing_key, _to_bytes(string_to_sign), hashlib.sha256).hexdigest()

    return "AWS4-HMAC-SHA256 Credential=%s/%s, SignedHeaders=%s, Signature=%s" % (
        cfg.access_key, scope, signed_headers, signature)


def _request(cfg, method, bucket, key=None, query="", body=None, extra_header=None):
    parts = urlsplit(cfg.endpoint)
    host = parts.netloc
    hostname = parts.hostname
    port = parts.port or (443 if parts.scheme == "https" else 80)
    canonical_uri = "/%s/%s" % (bucket, key) if key else "/%s" % bucket
    path = canonical_uri + (("?" + query) if query else "")

    now = time.gmtime()
    amzdate = time.strftime("%Y%m%dT%H%M%SZ", now)
    datestamp = time.strftime("%Y%m%d", now)
    body_bytes = _to_bytes(body) if body is not None else None
    content_length = len(body_bytes) if body_bytes is not None else None
    authorization = _sign_request(cfg, method, canonical_uri, query, host, amzdate, datestamp, content_length)

    headers = [("Host", host), ("x-amz-date", amzdate), ("Authorization", authorization)]
    if content_length is not None:
        headers.append(("Content-Length", str(content_length)))
    elif method == "PUT":
        headers.append(("Content-Length", "0"))
    if extra_header:
        headers.append(extra_header)

    conn_cls = http_client.HTTPSConnection if parts.scheme == "https" else http_client.HTTPConnection
    conn = conn_cls(hostname, port)
    try:
        conn.putrequest(method, path, skip_host=True, skip_accept_encoding=True)
        for k, v in headers:
            conn.putheader(k, v)
        conn.endheaders()
        if body_bytes is not None and body_bytes:
            conn.send(body_bytes)
        resp = conn.getresponse()
        data = resp.read()
        if not isinstance(data, str):
            data = data.decode("utf-8", "replace")
        if resp.status < 200 or resp.status >= 300:
            raise RuntimeError("%s failed: HTTP %d body=%s" % (method, resp.status, data))
        return data
    finally:
        conn.close()


def ws_url_from_endpoint(endpoint):
    parts = urlsplit(endpoint)
    scheme = "wss" if parts.scheme == "https" else "ws"
    return "%s://%s/rdma-ctrl" % (scheme, parts.netloc)


def _text(root, name):
    node = root.find(".//" + name)
    return node.text if node is not None and node.text is not None else ""


def create_multipart(cfg, bucket, obj):
    xml = _request(cfg, "POST", bucket, obj, "uploads=")
    upload_id = _text(ET.fromstring(xml), "UploadId")
    if not upload_id:
        raise RuntimeError("CreateMultipartUpload response missing UploadId: %s" % xml)
    print("Successfully created multipart upload")
    print("Upload ID: %s" % upload_id)
    return upload_id


def upload_part_rdma(cfg, args, upload_id, part_number, data):
    request_id = "rdma-multipart-%d-part-%d-%d" % (os.getpid(), part_number, int(time.time() * 1000000))
    sess = rdmalib.RDMASession.send_setup(ws_url_from_endpoint(args.endpoint), args.rdma_dev,
                                           rdmalib.RDMA_PORT, rdmalib.RDMA_GID_IDX, request_id, data)
    try:
        query = "partNumber=%d&uploadId=%s" % (part_number, upload_id)
        _request(cfg, "PUT", args.bucket, args.object, query, None, ("X-RDMA-Request-ID", request_id))
    finally:
        sess.destroy()
    etag = '"part-%d"' % part_number
    print("Uploaded part %d, size=%d, ETag=%s" % (part_number, len(data), etag), file=sys.stderr)
    return {"PartNumber": part_number, "ETag": etag, "Size": len(data)}


def upload_parts(cfg, args, upload_id):
    size = os.path.getsize(args.file)
    if size <= 0:
        raise RuntimeError("multipart RDMA upload does not support empty files")
    num_parts = int((size + args.partsize - 1) // args.partsize)
    if args.part < 0 or args.part > num_parts:
        raise RuntimeError("part must be between 1 and %d, or 0 for all parts" % num_parts)
    start = args.part or 1
    end = args.part or num_parts
    parts = []
    with open(args.file, "rb") as f:
        for part_number in range(start, end + 1):
            f.seek((part_number - 1) * args.partsize)
            data = f.read(args.partsize)
            parts.append(upload_part_rdma(cfg, args, upload_id, part_number, data))
    return parts


def list_parts(cfg, bucket, obj, upload_id):
    xml = _request(cfg, "GET", bucket, obj, "uploadId=%s" % upload_id)
    root = ET.fromstring(xml)
    parts = []
    for node in root.findall(".//Part"):
        parts.append({
            "PartNumber": int(_text(node, "PartNumber")),
            "ETag": _text(node, "ETag"),
            "Size": int(_text(node, "Size") or 0),
        })
    return parts


def print_parts(parts, upload_id):
    print("Parts uploaded for upload ID %s:" % upload_id)
    for p in parts:
        print("Part %d: ETag=%s, Size=%d bytes" % (p["PartNumber"], p["ETag"], p.get("Size", 0)))


def complete_multipart(cfg, bucket, obj, upload_id, parts):
    body = ["<CompleteMultipartUpload>"]
    for p in sorted(parts, key=lambda x: x["PartNumber"]):
        body.append("<Part><PartNumber>%d</PartNumber><ETag>%s</ETag></Part>" % (p["PartNumber"], p["ETag"]))
    body.append("</CompleteMultipartUpload>")
    xml = _request(cfg, "POST", bucket, obj, "uploadId=%s" % upload_id, "".join(body))
    print(xml)
    print("Successfully completed multipart upload")


def abort_multipart(cfg, bucket, obj, upload_id):
    _request(cfg, "DELETE", bucket, obj, "uploadId=%s" % upload_id)
    print("Successfully aborted multipart upload")


def list_uploads(cfg, args):
    qs = ["uploads="]
    if args.delimiter:
        qs.append("delimiter=%s" % args.delimiter)
    if args.key_marker:
        qs.append("key-marker=%s" % args.key_marker)
    if args.max_uploads:
        qs.append("max-uploads=%d" % args.max_uploads)
    if args.prefix:
        qs.append("prefix=%s" % args.prefix)
    if args.upload_id_marker:
        qs.append("upload-id-marker=%s" % args.upload_id_marker)
    print(_request(cfg, "GET", args.bucket, None, "&".join(qs)))


def main():
    p = argparse.ArgumentParser()
    p.add_argument("-action", default="all", choices=["all", "create", "upload", "list", "list-uploads", "complete", "abort"])
    p.add_argument("-bucket", required=True)
    p.add_argument("-object")
    p.add_argument("-file")
    p.add_argument("-uploadid")
    p.add_argument("-part", type=int, default=0)
    p.add_argument("-partsize", type=int, default=5 * 1024 * 1024)
    p.add_argument("-endpoint", default="http://127.0.0.1:8901")
    p.add_argument("-rdma-dev", dest="rdma_dev", default="rxe0")
    p.add_argument("-ak", default="admin-ak")
    p.add_argument("-sk", default="admin-sk")
    p.add_argument("-prefix", default="")
    p.add_argument("-delimiter", default="")
    p.add_argument("-max-uploads", dest="max_uploads", type=int, default=1000)
    p.add_argument("-key-marker", dest="key_marker", default="")
    p.add_argument("-upload-id-marker", dest="upload_id_marker", default="")
    args = p.parse_args()

    if args.action != "list-uploads" and not args.object:
        p.error("-object is required for this action")
    cfg = S3Config(args.endpoint, "us-east-1", args.ak, args.sk)

    if args.action == "create":
        create_multipart(cfg, args.bucket, args.object)
    elif args.action == "upload":
        if not args.file or not args.uploadid:
            p.error("-action upload requires -file and -uploadid")
        for part in upload_parts(cfg, args, args.uploadid):
            print("Part %d ETag: %s" % (part["PartNumber"], part["ETag"]))
    elif args.action == "list":
        if not args.uploadid:
            p.error("-action list requires -uploadid")
        print_parts(list_parts(cfg, args.bucket, args.object, args.uploadid), args.uploadid)
    elif args.action == "list-uploads":
        list_uploads(cfg, args)
    elif args.action == "complete":
        if not args.uploadid:
            p.error("-action complete requires -uploadid")
        parts = list_parts(cfg, args.bucket, args.object, args.uploadid)
        if not parts:
            raise RuntimeError("no uploaded parts found for uploadId %s" % args.uploadid)
        complete_multipart(cfg, args.bucket, args.object, args.uploadid, parts)
    elif args.action == "abort":
        if not args.uploadid:
            p.error("-action abort requires -uploadid")
        abort_multipart(cfg, args.bucket, args.object, args.uploadid)
    elif args.action == "all":
        if not args.file:
            p.error("-action all requires -file")
        upload_id = create_multipart(cfg, args.bucket, args.object)
        try:
            parts = upload_parts(cfg, args, upload_id)
            complete_multipart(cfg, args.bucket, args.object, upload_id, parts)
        except Exception:
            abort_multipart(cfg, args.bucket, args.object, upload_id)
            raise
    return 0


if __name__ == "__main__":
    sys.exit(main())
